File: python/codeforces/missile_silos/missile_silos.py
from heapq import heappush, heappop

def i_in():
    return (map(int, input().split()))

nodes, edges, start = i_in()
start -= 1

graph = [[] for x in range(nodes)]
edge_list = []
shortest_path = [float('inf') for x in range(nodes)]

for x in range(edges):
    left, right, cost = i_in()
    left -= 1; right -= 1
    graph[left].append((right, cost))
    graph[right].append((left, cost))
    edge_list.append((left, right, cost))

l, = i_in()

pq = [(0, start)]
while pq:
    path_cost, node = heappop(pq)
    if shortest_path[node] < path_cost: continue
    shortest_path[node] = path_cost
    for adj, cost in graph[node]:
        if cost + path_cost < shortest_path[adj]:
            shortest_path[adj] = cost + path_cost
            heappush(pq, (shortest_path[adj], adj))

silos = 0
for p in shortest_path:
    silos += 1 if p == l else 0

for left, right, cost in edge_list:
    if shortest_path[left] < l and shortest_path[left] + cost > l:
        silo_dist = l - shortest_path[left]
        silos += 1 if shortest_path[right] + (cost - silo_dist) >= shortest_path[left] + silo_dist else 0
    if shortest_path[right] < l and shortest_path[right] + cost > l:
        silo_dist = l - shortest_path[right]
        silos += 1 if shortest_path[left] + (cost - silo_dist) > shortest_path[right] + silo_dist else 0
print(silos)

File: cpp/codeforces/missile_silos/missile_silos.cpp
#include <iostream>
#include <tuple>
#include <queue>
#include <vector>
#include <algorithm>
#include <functional>
#include <climits>
using namespace std;

template<typename T> using min_pq = priority_queue<T, vector<T>, greater<T>>;

int main() {
    int nodes, edges, start;
    cin >> nodes >> edges >> start;
    start--;
    
    vector<vector<pair<int, int>>> graph(nodes);
    vector<int> shortestpath(nodes, INT_MAX >> 1);
    vector<tuple<int, int, int>> edge_list(edges);

    for(int i = 0; i < edges; i++) {
        int left, right, cost;
        cin >> left >> right >> cost;
        left--; right--;
        graph[left].push_back({right, cost});
        graph[right].push_back({left, cost});
        edge_list[i] = {left, right, cost};
    }
    int l;
    cin >> l;

    min_pq<pair<int, int>> pq;
    pq.push({0, start});
    
    while(!pq.empty()) {
        int path_cost, node; tie(path_cost, node) = pq.top(); pq.pop();
        if(shortestpath[node] < path_cost) continue;
        shortestpath[node] = path_cost;
        for(auto &e : graph[node]) {
            int adj = e.first, cost = e.second;
            if(shortestpath[adj] > path_cost + cost) {
                pq.push({path_cost + cost, adj});
            }
        }
    }

    int silos = 0;
    for(int &i : shortestpath) {
        silos += i==l ? 1:0;
    }

    for(auto &e : edge_list) {
        int left, right, cost; tie(left, right, cost) = e;
        if(shortestpath[left] < l && shortestpath[left] + cost > l) {
            int silo_dist = l - shortestpath[left];
            if(shortestpath[right] + cost - silo_dist >= l) {
                silos++;
            }
        }
        if(shortestpath[right] < l && shortestpath[right] + cost > l) {
            int silo_dist = l - shortestpath[right];
            if(shortestpath[left] + cost - silo_dist > l) {
                silos++;
            }
        }
    }
    cout << silos;
}
